jniFaiss/src/com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.cpp

```diff
@@ -0,0 +1,272 @@
+#include "com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex.h"
+
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <string>
+#include <vector>
+#include <sys/time.h>
+#include <omp.h>
+
+#include "faiss/index_factory.h"
+#include "faiss/MetaIndexes.h"
+#include "faiss/index_io.h"
+#include "faiss/IndexHNSW.h"
+
+
+using std::string;
+using std::vector;
+
+std::unordered_map<string, faiss::MetricType> mapMetric = {
+        {"l2", faiss::METRIC_L2},
+        {"innerproduct", faiss::METRIC_INNER_PRODUCT}
+};
+
+extern "C"
+
+struct JavaException {
+    JavaException(JNIEnv* env, const char* type = "", const char* message = "")
+    {
+        jclass newExcCls = env->FindClass(type);
+        if (newExcCls != NULL)
+            env->ThrowNew(newExcCls, message);
+    }
+};
+
+inline void has_exception_in_stack(JNIEnv* env)
+{
+    if (env->ExceptionCheck() == JNI_TRUE)
+        throw std::runtime_error("Exception Occured");
+}
+
+void catch_cpp_exception_and_throw_java(JNIEnv* env)
+{
+    try {
+        throw;
+    }
+    catch (const std::bad_alloc& rhs) {
+        JavaException(env, "java/io/IOException", rhs.what());
+    }
+    catch (const std::runtime_error& re) {
+        JavaException(env, "java/lang/Exception", re.what());
+    }
+    catch (const std::exception& e) {
+        JavaException(env, "java/lang/Exception", e.what());
+    }
+    catch (...) {
+        JavaException(env, "java/lang/Exception", "Unknown exception occured");
+    }
+}
+
+/**
+ * Method: saveIndex
+ *
+ */
+JNIEXPORT void JNICALL Java_com_amazon_opendistroforelasticsearch_knn_index_faiss_v164_KNNFIndex_saveIndex
+(JNIEnv* env, jclass cls, jintArray ids, jobjectArray vectors, jstring indexPath, jobjectArray algoParams, jstring spaceType)
+{
+	vector<int64_t> idVector;
+	vector<float>   dataset;
+	vector<string> paramsList;
+	string indexDescription = "HNSW32";
+	faiss::MetricType metric = faiss::METRIC_L2;
```

Comment:
I think instead of setting L2 as default, we should just make sure that spaceType is passed in and is valid. Setting L2 as default here could lead to silent failures.

Reply:
L72 just init a default value. at L121 i verify the matric:
```
if(mapMetric.find(spaceTypeString) != mapMetric.end()) {
    metric = mapMetric[spaceTypeString];
}
```

I have a question: if we verify the `metricSpace` in the jni code. this may cause ES shard failed assert when an invalid space param passed in jni code.  should i throw an exception when mapMetric can not find spaceTypeString?